crates/ruff_python_semantic/src/red_knot/types/infer.rs

```diff
@@ -0,0 +1,980 @@
+            ast::Expr::Named(named) => self.lower_named_expression(named),
+            ast::Expr::If(if_expression) => self.lower_if_expression(if_expression),
+
+            _ => todo!("expression type resolution for {:?}", expression),
+        };
+
+        self.types.expression_tys.push(ty);
+
+        ty
+    }
+
+    #[allow(clippy::unused_self)]
+    fn lower_number_literal_expression(&mut self, literal: &ast::ExprNumberLiteral) -> Type {
+        let ast::ExprNumberLiteral { range: _, value } = literal;
+
+        match value {
+            ast::Number::Int(n) => {
+                // TODO support big int literals
+                n.as_i64().map(Type::IntLiteral).unwrap_or(Type::Unknown)
+            }
+            // TODO builtins.float or builtins.complex
+            _ => Type::Unknown,
+        }
+    }
+
+    fn lower_named_expression(&mut self, named: &ast::ExprNamed) -> Type {
+        let ast::ExprNamed {
+            range: _,
+            target,
+            value,
+        } = named;
+
+        let value_ty = self.lower_expression(value);
+        self.lower_expression(target);
+
+        self.definition_tys.insert(
+            Definition::NamedExpr(named.scope_ast_id(self.db, self.file_id, self.file_scope_id)),
+            value_ty,
+        );
+
+        value_ty
+    }
+
+    fn lower_if_expression(&mut self, if_expression: &ast::ExprIf) -> Type {
+        let ast::ExprIf {
+            range: _,
+            test,
+            body,
+            orelse,
+        } = if_expression;
+
+        self.lower_expression(test);
+
+        // TODO detect statically known truthy or falsy test
+        let body_ty = self.lower_expression(body);
+        let orelse_ty = self.lower_expression(orelse);
+
+        let union = UnionTypeBuilder::new(&self.typing_context())
+            .variant(body_ty)
+            .variant(orelse_ty)
+            .build();
+
+        self.union_ty(union)
+    }
+
+    fn lower_name_expression(&mut self, name: &ast::ExprName) -> Type {
+        let ast::ExprName { range: _, id, ctx } = name;
+
+        match ctx {
+            ExprContext::Load => {
+                // TODO Consider non local scopes?
+                if let Some(symbol_id) = self
+                    .index
+                    .symbol_table(self.file_scope_id)
+                    .symbol_id_by_name(id)
+                {
+                    self.local_definition_ty(symbol_id).into_type(self)
+                } else {
+                    let ancestors = self.index.ancestor_scopes(self.file_scope_id).skip(1);
+
+                    for (ancestor_id, _) in ancestors {
+                        let symbol_table =
+                            symbol_table(self.db, ancestor_id.to_scope_id(self.db, self.file_id));
+
+                        if let Some(_symbol_id) = symbol_table.symbol_id_by_name(id) {
+                            todo!("Return type for symbol from outer scope");
+                        }
```

Comment:
>  symbols in class scopes are never visible from nested scopes, unless they are immediately-nested type param scopes
Good point. 

> we are missing the fallback to builtins scope

Maybe, it depends on how we model builtin scopes. Is a builtin scope just a scope that we add to every module and is part of `scopes`? In that case, builtins would be handled already ;)

Reply:
Yeah, good point, that way of handling builtin scope might work out fine. Though it would save a bit of memory to just special-case it in lookup instead of adding an extra entry to every scopes array that is always the same. (Similar to the discussion of which way to handle `Definition::Unbound`.)